### PR TITLE
Implement support for transforms that apply to every type and use compaction for aging

### PR DESCRIPTION
**Pull request by lava**
Aging used to work only partially; deleting events from the archive but not from the index. It was deprecated and scheduled for removal in VAST 2.0.

With this PR we change the internals of the eraser, so aging now works as a functionality-reduced form of temporal compaction.

Some supporting commits add support for transforms that apply to every type, and a new "invert" option for the "select" transform step.

### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### CODE REVIEW

File: libvast/include/vast/system/eraser.hpp [hunk 1]
```diff
@@ -12,67 +12,55 @@
...
 };
 
 /// Periodically queries `index` and erases all hits from `archive`.
 /// @param interval The time between two query executions.
 /// @param query The periodic query that selects events scheduled for deletion.
 ///              Note that we get the query as string on purpose. Taking an
-///              ::expression here instead would fix any query such as `#time <
-///              1 week ago` to the time of its parsing and not update
+///              ::expression here instead would fix any query such as
+///              `#time < 1 week ago` to the time of its parsing and not update
```
Comment (dominiklohmann) [inline libvast/include/vast/system/eraser.hpp:58]: Nit: `#time < 1 week ago` is not a valid query. Same above in the documentation of the `query_` member variable of the state.

File: libvast/include/vast/system/eraser.hpp [hunk 2]
```diff
@@ -12,67 +12,55 @@
...
 ///              properly.
 /// @param index A handle to the INDEX under investigation.
-caf::behavior
-eraser(caf::stateful_actor<eraser_state>* self, caf::timespan interval,
-       std::string query, index_actor index);
+/// @param transforms The transform config of the node.
```
Comment (dominiklohmann) [inline libvast/include/vast/system/eraser.hpp:61]: This parameter was not added to the actual function signature afaict.

File: libvast/include/vast/system/make_transforms.hpp [hunk 3]
```diff
@@ -38,7 +38,7 @@ make_transforms(transforms_location location, const caf::settings& settings);
 /// name.
 caf::expected<transform_ptr>
 make_transform(const std::string& name,
-               const std::vector<std::string>& event_types,
+               const std::optional<std::vector<std::string>>& event_types,
```
Comment (dominiklohmann) [inline libvast/include/vast/system/make_transforms.hpp:41]: Isn't the optional part already covered by the empty vector?

File: libvast/include/vast/transform.hpp [hunk 4]
```diff
@@ -33,18 +34,22 @@ class transform {
...
   /// @note The offsets of the slices may not be preserved.
   [[nodiscard]] caf::expected<std::vector<table_slice>> finish();
 
-  [[nodiscard]] const std::vector<std::string>& event_types() const;
-
   [[nodiscard]] const std::string& name() const;
 
 private:
+  [[nodiscard]] const std::optional<std::vector<std::string>>&
+  event_types() const;
```
Comment (dominiklohmann) [inline libvast/include/vast/transform.hpp:51]: Should probably be named "schema_names" or so.

File: libvast/include/vast/transform_steps/select.hpp [hunk 5]
```diff
@@ -18,22 +18,26 @@ struct select_step_configuration {
   // The expression in the config file.
   std::string expression;
 
+  // Whether to select or to filter.
+  bool invert = false;
```
Comment (dominiklohmann) [inline libvast/include/vast/transform_steps/select.hpp:22]: We have both `project` and `delete` steps, so why not introduce a `filter` step whose plugin implementation simply refers to the `filter` plugin?

File: libvast/src/system/eraser.cpp [hunk 6]
```diff
@@ -14,80 +14,102 @@
...
+        = std::make_shared<vast::transform>("eraser_transform", std::nullopt);
+      auto select_config = select_step_configuration{
+        .expression = self->state.query_,
+        .invert = true,
+      };
+      transform->add_step(std::make_unique<select_step>(select_config));
+      auto rp = self->make_response_promise<atom::ok>();
+      self->request(self->state.index_, caf::infinite, atom::resolve_v, *expr)
+        .then(
+          [self, transform, rp = rp](catalog_result& result) mutable {
```
Comment (dominiklohmann) [inline libvast/src/system/eraser.cpp:84]: Why the `rp = rp` over just `rp` here and a few times below?

**Review by dominiklohmann — COMMENTED** [review]
This is great, and thanks especially for turning the eraser into a typed actor.

I think this is almost ready to go. We need to undeprecate aging and also mark it as no longer experimental, and the rest should be covered by the remaining TODO comments and the few comments I left on the PR, which are mostly of stylistic nature.

File: libvast/include/vast/system/make_transforms.hpp (hunk 3 above)
Comment (lava) [reply libvast/include/vast/system/make_transforms.hpp:41]: Good question, I thought about it but felt that using an explicit optional is less surprising. I don't have a strong opinion though, so I'll change to empty vector.

**Review by dominiklohmann — APPROVED** [review]

File: libvast/src/system/eraser.cpp (hunk 6 above)
Comment (lava) [reply libvast/src/system/eraser.cpp]: Good question, there used to be a mysterious compiler error if I didn't, but I can't reproduce.

**Comment by dominiklohmann** [comment]
This is only missing a rebase, a changelog entry and a documentation update now, everything else is now good to go.

**Review by dispanser — APPROVED** [review]
I verified the last four commits done by @dominiklohmann . We talked about those earlier.

**Comment by dominiklohmann** [comment]
[sc-32123]

**Pull request merged**